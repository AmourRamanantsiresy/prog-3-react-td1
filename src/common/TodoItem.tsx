import {
  TI_INPUT_TODO_DELETE,
  TI_INPUT_TODO_DONE,
  TR_INPUT_TODO_DELETE,
  TR_INPUT_TODO_DONE,
} from '../tests/utils/selectors';
import Todo, { TodoStatus } from '../utils/models/Todo';

type TodoItemProps = {
  todo: Todo;
  onDone?: (todoId: number) => void;
  onRemove?: (todoId: number) => void;
};

const TodoItem = (props: TodoItemProps) => {
  const { onDone, todo, onRemove } = props;

  const toDone = () => {
    onDone && onDone(todo.id);
  };

  const remove = () => {
    onRemove && onRemove(todo.id);
  };

  return (
    <div className='flex todo-item'>
      {todo.status === TodoStatus.TODO && (
        <input
          type='checkbox'
          className='pointer'
          onChange={toDone}
          id={todo.id.toString()}
          // tests props
          data-testid={`${TI_INPUT_TODO_DONE}-${todo.id}`}
          role={TR_INPUT_TODO_DONE}
        />
      )}
      <label className='ml-2'>{todo.value}</label>
      {todo.status === TodoStatus.DONE && (
        <p
          className='ml-2 pointer'
          onClick={remove}
          // tests props
          data-testid={`${TI_INPUT_TODO_DELETE}-${todo.id}`}
          role={TR_INPUT_TODO_DELETE}
        >
          x
        </p>
      )}
    </div>
  );
};

export default TodoItem;
