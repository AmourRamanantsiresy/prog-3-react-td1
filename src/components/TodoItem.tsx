import { ID_TODO_DELETE, ID_TODO_DONE, ROLE_TODO_DELETE } from '../tests/utils/selectors';
import Todo, { TodoStatus } from '../utils/models/Todo';

type TodoItemProps = {
  todo: Todo;
  onDone?: (todoId: number) => void;
  onRemove?: (todoId: number) => void;
};

const TodoItem = (props: TodoItemProps) => {
  const { onDone, todo, onRemove } = props;

  const toDone = () => {
    onDone && onDone(todo.id);
  };

  const remove = () => {
    onRemove && onRemove(todo.id);
  };

  return (
    <div className='flex todo-item'>
      {todo.status === TodoStatus.TODO && (
        <input
          type='checkbox'
          className='pointer'
          onChange={toDone}
          id={todo.id.toString()}
          // tests props
          data-testid={`${ID_TODO_DONE}-${todo.id}`}
        />
      )}
      <label className='ml-2'>{todo.value}</label>
      {todo.status === TodoStatus.DONE && (
        <p
          className='ml-2 pointer'
          onClick={remove}
          // tests props
          data-testid={`${ID_TODO_DELETE}-${todo.id}`}
          role={ROLE_TODO_DELETE}
        >
          x
        </p>
      )}
    </div>
  );
};

export default TodoItem;
