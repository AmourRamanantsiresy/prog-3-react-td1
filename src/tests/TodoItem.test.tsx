import { fireEvent, render, screen } from '@testing-library/react';
import { create, ReactTestRendererJSON } from 'react-test-renderer';
import TodoItem from '../components/TodoItem';
import { TODO_1_MOCK, TODO_DONE_1_MOCK } from './utils/mocks/todo-mock';
import { ID_TODO_DELETE, ID_TODO_DONE, ROLE_TODO_DELETE } from './utils/selectors';

describe('Test TodoItem', () => {
  test('Architecture', () => {
    const tree = create(
      <TodoItem
        key='todo_1'
        onDone={todoId => {
          console.log(todoId);
        }}
        onRemove={todoId => {
          console.log(todoId);
        }}
        todo={TODO_1_MOCK}
      />
    ).toJSON() as ReactTestRendererJSON;

    expect(tree).toMatchSnapshot();
  });

  test('When todo status === TODO', () => {
    const toDoneSelector = `${ID_TODO_DONE}-${TODO_1_MOCK.id}`;
    const handleDone: jest.Mock<number> = jest.fn(id => id);
    render(<TodoItem key='todo_1' onDone={handleDone} todo={TODO_1_MOCK} />);

    screen.getByText(TODO_1_MOCK.value);
    expect(screen.getByTestId(toDoneSelector)).not.toBeChecked();
    expect(screen.queryByRole(ROLE_TODO_DELETE)).toBeNull();
    fireEvent.click(screen.getByTestId(toDoneSelector));
    expect(handleDone.mock.results[0].value).toEqual(TODO_1_MOCK.id);
  });

  test('When todo status === DONE', () => {
    const deleteTodoSelector = `${ID_TODO_DELETE}-${TODO_DONE_1_MOCK.id}`;
    const handleRemove: jest.Mock<number> = jest.fn(id => id);
    render(<TodoItem key='todo_1' onRemove={handleRemove} todo={TODO_DONE_1_MOCK} />);

    screen.getByText(TODO_DONE_1_MOCK.value);
    screen.getByText('x');

    fireEvent.click(screen.getByTestId(deleteTodoSelector));

    expect(handleRemove.mock.results[0].value).toEqual(TODO_DONE_1_MOCK.id);
    expect(screen.queryByRole('checkbox')).toBeNull();
    expect(screen.queryByTestId(deleteTodoSelector)).toBeTruthy();
  });
});
