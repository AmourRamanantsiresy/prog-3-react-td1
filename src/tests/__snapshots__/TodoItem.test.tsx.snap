// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoItem Architecture 1`] = `
<div
  className="flex todo-item"
>
  <input
    className="pointer"
    data-testid="input-todo-done-id-1"
    id="1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="ml-2"
  >
    Todo number 1
  </label>
</div>
`;
